Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp

```diff
@@ -0,0 +1,184 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+
+#include "UserExperienceComponent.h"
+
+#include "Net/UnrealNetwork.h"
+#include "SpatialNetDriver.h"
+#include "Utils/SpatialMetrics.h"
+
+DEFINE_LOG_CATEGORY(LogUserExperienceComponent);
+
+namespace
+{
+	float Calculate80thPctAverage(const TArray<float>& Array)
+	{
+		TArray<float> Sorted = Array;
+		Sorted.Sort();
+		int ElementsToAverage = Sorted.Num()*0.8f; // Exclude 20% of samples
+
+		float Avg = 0.0f;
+		for (int i = 0; i < ElementsToAverage; i++)
+		{
+			Avg += Sorted[i];
+		}
+		Avg /= ElementsToAverage;
+		return Avg;
+	}
+}
+
+// Sets default values for this component's properties
+UUserExperienceComponent::UUserExperienceComponent()
+{
+	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
+	// off to improve performance if you don't need them.
+	PrimaryComponentTick.bCanEverTick = true;
+	bWantsInitializeComponent = true;
+	bAutoActivate = true;
+}
+
+void UUserExperienceComponent::InitializeComponent()
+{
+	ElapsedTime = 0.0f;
+	ServerClientRTT = 0.0f;
+	ServerViewLateness = 0.0f;
+	ClientRTTTimer = 0.0f; 
+	SetIsReplicated(true);
+	UActorComponent::InitializeComponent();
+}
+
+void UUserExperienceComponent::SendServerRPC()
+{
+	ServerRTT(ElapsedTime);
+}
+
+void UUserExperienceComponent::UpdateClientObservations(float DeltaTime)
```

Comment:
What if we used an OnRep callback here instead?

Reply:
I left this the way it was because it looks at the replicated values of other UUserExpereinceComponents so it's important that all replication has already occurred